File: orchestrator-tui/TokenManager.cs
using System.Net;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Net;
// ... (using lain) ...
using Spectre.Console;

namespace Orchestrator;

public static class TokenManager
{
    // ... (Path variables, _tokens, _availableProxies, _state, _tokenCache tetap sama) ...
    private static readonly string ConfigRoot = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory!, "..", "..", "..", "..", "config"));
    private static readonly string ProjectRoot = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory!, "..", "..", "..", ".."));
    private static readonly string TokensPath = Path.Combine(ConfigRoot, "github_tokens.txt");
    private static readonly string SuccessProxyListPath = Path.Combine(ProjectRoot, "proxysync", "success_proxy.txt");
    private static readonly string FallbackProxyListPath = Path.Combine(ProjectRoot, "proxysync", "proxy.txt");
    private static readonly string StatePath = Path.Combine(ProjectRoot, ".token-state.json");
    private static readonly string TokenCachePath = Path.Combine(ProjectRoot, ".token-cache.json");

    private static List<TokenEntry> _tokens = new();
    private static List<string> _availableProxies = new();
    private static TokenState _state = new();
    private static Dictionary<string, string> _tokenCache = new();

    private const int DEFAULT_HTTP_TIMEOUT_SEC = 60;

    // Initialize tetap panggil semua load
    public static void Initialize() { LoadTokens(); LoadProxyList(); LoadState(); LoadTokenCache(); AssignProxiesAndUsernames(); }

    // ReloadAllConfigs HARUS reset state juga (buat kasus config token berubah total)
    public static void ReloadAllConfigs() {
        AnsiConsole.MarkupLine("[yellow]Reloading ALL configs (Tokens, Proxies, State)...[/]");
        _tokens.Clear();
        _availableProxies.Clear();
        _tokenCache.Clear(); // Cache juga perlu di-clear kalo token berubah
        _state = new TokenState(); // Reset state ke awal
        Initialize(); // Panggil ulang semua load
        AnsiConsole.MarkupLine("[green]All configs refreshed.[/]");
    }

    // === FUNGSI BARU: Hanya reload proxy & reassign ===
    public static void ReloadProxyListAndReassign() {
        AnsiConsole.MarkupLine("[yellow]Reloading proxy list and reassigning...[/]");
        _availableProxies.Clear(); // Kosongkan list proxy lama
        LoadProxyList(); // Baca ulang success_proxy.txt / proxy.txt
        AssignProxiesToExistingTokens(); // Assign ulang proxy ke token yang sudah ada
        AnsiConsole.MarkupLine("[green]Proxy list refreshed and reassigned.[/]");
    }
    // === AKHIR FUNGSI BARU ===

    // Helper baru untuk assign proxy tanpa load token ulang
    private static void AssignProxiesToExistingTokens() {
        if (!_tokens.Any()) return;
        AnsiConsole.MarkupLine($"[dim]   Assigning {_availableProxies.Count} proxies to {_tokens.Count} existing tokens...[/]");
        for (int i = 0; i < _tokens.Count; i++) {
            if (_availableProxies.Any()) {
                _tokens[i].Proxy = _availableProxies[i % _availableProxies.Count];
            } else {
                _tokens[i].Proxy = null; // Hapus proxy jika list baru kosong
            }
        }
        // Tidak perlu panggil InitializeProxyPool karena pool lama udah dihapus
    }


    // ... (LoadTokenCache, SaveTokenCache, LoadTokens, LoadProxyList tetap sama) ...
     private static void LoadTokenCache() { if (!File.Exists(TokenCachePath)) return; try { var json = File.ReadAllText(TokenCachePath); _tokenCache = JsonSerializer.Deserialize<Dictionary<string, string>>(json) ?? new(); AnsiConsole.MarkupLine($"[dim]Loaded {_tokenCache.Count} usernames from cache[/]"); } catch (Exception ex) { AnsiConsole.MarkupLine($"[yellow]Warn: Load cache fail: {ex.Message.EscapeMarkup()}[/]"); _tokenCache = new(); } }
    public static void SaveTokenCache(Dictionary<string, string> cache) { try { _tokenCache = cache; var json = JsonSerializer.Serialize(_tokenCache, new JsonSerializerOptions { WriteIndented = true }); File.WriteAllText(TokenCachePath, json); } catch (Exception ex) { AnsiConsole.MarkupLine($"[red]Err save cache: {ex.Message.EscapeMarkup()}[/]"); } }
    private static void LoadTokens() { if (!File.Exists(TokensPath)) { AnsiConsole.MarkupLine($"[red]ERR: {TokensPath.EscapeMarkup()} not found![/]"); return; } var lines = File.ReadAllLines(TokensPath); if (lines.Length < 3) { AnsiConsole.MarkupLine("[red]ERR: github_tokens.txt format incorrect![/]"); return; } var owner = lines[0].Trim(); var repo = lines[1].Trim(); var tokens = lines[2].Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries); _tokens = tokens.Select(t => new TokenEntry { Token = t, Owner = owner, Repo = repo }).ToList(); AnsiConsole.MarkupLine($"[green]Loaded {_tokens.Count} tokens for {owner.EscapeMarkup()}/{repo.EscapeMarkup()}[/]"); }
    private static void LoadProxyList() { string proxyFileToLoad = File.Exists(SuccessProxyListPath) && new FileInfo(SuccessProxyListPath).Length > 0 ? SuccessProxyListPath : File.Exists(FallbackProxyListPath) && new FileInfo(FallbackProxyListPath).Length > 0 ? FallbackProxyListPath : ""; if (string.IsNullOrEmpty(proxyFileToLoad)) { AnsiConsole.MarkupLine("[red]CRITICAL: No proxy file found or proxy files are empty![/]"); AnsiConsole.MarkupLine("[yellow]Run ProxySync first![/]"); _availableProxies.Clear(); return; } try { _availableProxies = File.ReadAllLines(proxyFileToLoad).Select(l => l.Trim()).Where(l => !string.IsNullOrWhiteSpace(l) && !l.StartsWith("#")).Distinct().ToList(); if (!_availableProxies.Any()) { AnsiConsole.MarkupLine($"[yellow]Warning: Proxy file '{Path.GetFileName(proxyFileToLoad)}' loaded but contains no valid proxies.[/]"); } else if (proxyFileToLoad == SuccessProxyListPath) AnsiConsole.MarkupLine($"[green]✓ {_availableProxies.Count} tested proxies loaded from success_proxy.txt[/]"); else { AnsiConsole.MarkupLine($"[yellow]⚠ Using fallback proxy.txt ({_availableProxies.Count} proxies)[/]"); AnsiConsole.MarkupLine("[yellow]  Consider running ProxySync test for better reliability![/]"); } } catch (Exception ex) { AnsiConsole.MarkupLine($"[red]Error loading proxies from '{Path.GetFileName(proxyFileToLoad)}': {ex.Message.EscapeMarkup()}[/]"); _availableProxies.Clear(); } }


    // AssignProxiesAndUsernames hanya dipanggil di Initialize awal
    private static void AssignProxiesAndUsernames() {
         if (!_tokens.Any()) return;
         AssignProxiesToExistingTokens(); // Panggil helper
         // Assign username dari cache
         foreach(var tokenEntry in _tokens) {
              if (_tokenCache.TryGetValue(tokenEntry.Token, out var u)) {
                   tokenEntry.Username = u;
              }
         }
     }


    // ... (LoadState, GetState, SaveState, GetCurrentToken, GetAllTokenEntries, GetUsernameCache tetap sama) ...
     private static void LoadState() { try { if (File.Exists(StatePath)) { var json = File.ReadAllText(StatePath); _state = JsonSerializer.Deserialize<TokenState>(json) ?? new TokenState(); AnsiConsole.MarkupLine($"[dim]Loaded state: Index={_state.CurrentIndex}, Codespace='{_state.ActiveCodespaceName ?? "None"}'[/]");} else { _state = new TokenState(); AnsiConsole.MarkupLine("[dim]No state file found, starting fresh.[/]");} } catch (Exception ex) { AnsiConsole.MarkupLine($"[red]Error loading state: {ex.Message.EscapeMarkup()}. Resetting state.[/]"); _state = new TokenState(); } if (_state.CurrentIndex >= _tokens.Count && _tokens.Any()) { AnsiConsole.MarkupLine($"[yellow]Warning: Saved index ({_state.CurrentIndex}) out of bounds ({_tokens.Count} tokens). Resetting to 0.[/]"); _state.CurrentIndex = 0; } }
    public static TokenState GetState() => _state;
    public static void SaveState(TokenState state) { _state = state; var json = JsonSerializer.Serialize(_state, new JsonSerializerOptions { WriteIndented = true }); try { File.WriteAllText(StatePath, json); } catch (Exception ex) { AnsiConsole.MarkupLine($"[red]Error saving state: {ex.Message.EscapeMarkup()}[/]"); } }
    public static TokenEntry GetCurrentToken() { if (!_tokens.Any()) throw new InvalidOperationException("No tokens configured in github_tokens.txt"); if (_state.CurrentIndex >= _tokens.Count || _state.CurrentIndex < 0) { AnsiConsole.MarkupLine($"[yellow]Warning: Current index ({_state.CurrentIndex}) is invalid. Resetting to 0.[/]"); _state.CurrentIndex = 0; SaveState(_state); } return _tokens[_state.CurrentIndex]; }
    public static List<TokenEntry> GetAllTokenEntries() => _tokens;
    public static Dictionary<string, string> GetUsernameCache() => _tokenCache;


    // Logika RotateProxyForToken tetap sama (versi pintar tanpa pool)
    public static bool RotateProxyForToken(TokenEntry currentTokenEntry)
    {
        if (!_availableProxies.Any()) { AnsiConsole.MarkupLine("[red]No proxies available to rotate.[/]"); return false; }
        string? oldProxy = currentTokenEntry.Proxy;
        var proxiesInUseByOthers = _tokens.Where(t => t.Token != currentTokenEntry.Token && !string.IsNullOrEmpty(t.Proxy)).Select(t => t.Proxy!).ToHashSet();
        string? newProxy = _availableProxies.FirstOrDefault(p => !proxiesInUseByOthers.Contains(p) && p != oldProxy);
        if (newProxy == null) {
            AnsiConsole.MarkupLine("[yellow]All proxies currently assigned. Picking random fallback (excluding current)...[/]");
            newProxy = _availableProxies.Where(p => p != oldProxy).OrderBy(x => Guid.NewGuid()).FirstOrDefault() ?? _availableProxies.FirstOrDefault(); // Fallback ke acak / yg pertama
        }
        if (newProxy == null) { AnsiConsole.MarkupLine("[red]FATAL: No proxies found in list after filtering.[/]"); return false; }
        AnsiConsole.MarkupLine($"[yellow]Proxy rotated: {MaskProxy(oldProxy)} -> {MaskProxy(newProxy)}[/]");
        currentTokenEntry.Proxy = newProxy;
        return true;
    }

    // Logika SwitchToNextToken tetap sama
    public static TokenEntry SwitchToNextToken() { if (!_tokens.Any() || _tokens.Count == 1) { AnsiConsole.MarkupLine("[yellow]Only 1 token configured or no tokens loaded.[/]"); return GetCurrentToken(); } _state.CurrentIndex = (_state.CurrentIndex + 1) % _tokens.Count; _state.ActiveCodespaceName = null; SaveState(_state); var current = _tokens[_state.CurrentIndex]; var username = current.Username ?? "unknown"; AnsiConsole.MarkupLine($"[yellow]Token Rotated: -> #{_state.CurrentIndex + 1} (@{username.EscapeMarkup()})[/]"); if (!string.IsNullOrEmpty(current.Proxy)) AnsiConsole.MarkupLine($"[dim]Proxy assigned: {MaskProxy(current.Proxy)}[/]"); return current; }

    // Logika CreateHttpClient tetap sama
     public static HttpClient CreateHttpClient(TokenEntry token) {
        var handler = new HttpClientHandler();
        if (!string.IsNullOrEmpty(token.Proxy)) {
            // AnsiConsole.Markup($"[dim]   Using proxy {MaskProxy(token.Proxy)}[/]"); // Kurangi verbosity
            try {
                var proxyUri = new Uri(token.Proxy, UriKind.Absolute);
                var webProxy = new WebProxy(proxyUri);
                if (!string.IsNullOrEmpty(proxyUri.UserInfo)) {
                    var credentials = proxyUri.UserInfo.Split(':', 2);
                    if (credentials.Length == 2) {
                        webProxy.Credentials = new NetworkCredential(Uri.UnescapeDataString(credentials[0]), Uri.UnescapeDataString(credentials[1]));
                        webProxy.UseDefaultCredentials = false;
                        // AnsiConsole.MarkupLine(" with credentials"); // Kurangi verbosity
                    } else { /* AnsiConsole.MarkupLine("[yellow] (Credentials format issue? Using proxy without auth)[/]"); */ webProxy.UseDefaultCredentials = true; }
                } else { webProxy.UseDefaultCredentials = true; /* AnsiConsole.MarkupLine(""); */ }
                handler.Proxy = webProxy;
                handler.UseProxy = true;
            } catch (UriFormatException ex) { AnsiConsole.MarkupLine($"[red]Proxy format invalid {MaskProxy(token.Proxy)}: {ex.Message.EscapeMarkup()}. Disabling proxy.[/]"); handler.Proxy = null; handler.UseProxy = false;
            } catch (Exception ex) { AnsiConsole.MarkupLine($"[red]Error setting proxy {MaskProxy(token.Proxy)}: {ex.Message.EscapeMarkup()}. Disabling proxy.[/]"); handler.Proxy = null; handler.UseProxy = false; }
        } else { handler.Proxy = null; handler.UseProxy = false; }
        var client = new HttpClient(handler) { Timeout = TimeSpan.FromSeconds(DEFAULT_HTTP_TIMEOUT_SEC) };
        client.DefaultRequestHeaders.Add("Authorization", $"Bearer {token.Token}");
        client.DefaultRequestHeaders.Add("User-Agent", "Automation-Hub-Orchestrator/4.1"); // Versi naik
        client.DefaultRequestHeaders.Add("Accept", "application/vnd.github+json");
        return client;
    }

    // MaskToken dan MaskProxy tetap sama
    public static string MaskToken(string token) { if (string.IsNullOrEmpty(token)) return "invalid"; return token.Length > 10 ? token[..4] + "..." + token[^4..] : token; }
    public static string MaskProxy(string? proxy) { if (string.IsNullOrEmpty(proxy)) return "[grey]no-proxy[/]"; try { if (Uri.TryCreate(proxy, UriKind.Absolute, out var uri)) { return $"{uri.Scheme}://{uri.Host}:{uri.Port}"; } var parts = proxy.Split('@'); if (parts.Length == 2) { var hostPort = parts[1].Split(':'); if (hostPort.Length >= 2) return $"proxy://{hostPort[0]}:{hostPort[1]}"; } else { var hostPort = proxy.Split(':'); if(hostPort.Length >= 2) return $"proxy://{hostPort[0]}:{hostPort[1]}"; } return "[grey]masked[/]"; } catch { return "[red]invalid-format[/]"; } }

    // ShowStatus tetap sama
     public static void ShowStatus() { if (!_tokens.Any()) { AnsiConsole.MarkupLine("[yellow]No tokens configured.[/]"); return; } var owner = _tokens.FirstOrDefault()?.Owner ?? "N/A"; var repo = _tokens.FirstOrDefault()?.Repo ?? "N/A"; var activeCs = _state.ActiveCodespaceName ?? "[grey]None[/]"; AnsiConsole.MarkupLine($"[cyan]Owner       :[/] [yellow]{owner.EscapeMarkup()}[/]"); AnsiConsole.MarkupLine($"[cyan]Repo        :[/] [yellow]{repo.EscapeMarkup()}[/]"); AnsiConsole.MarkupLine($"[cyan]Active CS   :[/] [yellow]{activeCs.EscapeMarkup()}[/]"); var table = new Table().Title("Tokens Status").Expand(); table.AddColumn("Idx"); table.AddColumn("Token"); table.AddColumn("User"); table.AddColumn("Proxy"); table.AddColumn("Active"); for (int i = 0; i < _tokens.Count; i++) { var t = _tokens[i]; var act = i == _state.CurrentIndex ? "[bold green]>>>[/]" : ""; var tokD = MaskToken(t.Token); var userD = t.Username ?? "[grey]unknown[/]"; var proxD = MaskProxy(t.Proxy); table.AddRow((i+1).ToString(), tokD, userD, proxD, act); } AnsiConsole.Write(table); AnsiConsole.MarkupLine($"[dim]Proxies available in list: {_availableProxies.Count}[/]"); }

} // Akhir class TokenManager

// Definisi class TokenEntry dan TokenState tetap sama
public class TokenEntry { public string Token { get; set; } = string.Empty; public string? Proxy { get; set; } public string? Username { get; set; } public string Owner { get; set; } = string.Empty; public string Repo { get; set; } = string.Empty; }
public class TokenState { [JsonPropertyName("current_index")] public int CurrentIndex { get; set; } = 0; [JsonPropertyName("active_codespace_name")] public string? ActiveCodespaceName { get; set; } = null; }

File: py-orchestrator/tokenmanager.py
import json
from pathlib import Path
from typing import List, Dict, Optional, Tuple, Any
from pydantic import BaseModel, Field
from datetime import datetime
import httpx
from rich.console import Console
from rich.table import Table

console = Console()

# Path
TOKENS_PATH = Path("../config/github_tokens.txt")
STATE_PATH = Path("../.token-state.json")
PROXY_LIST_PATH = Path("../proxysync/proxy.txt")
TOKEN_CACHE_PATH = Path("../.token-cache.json")


class BotExecutionState(BaseModel):
    last_step: int = Field(default=0, alias="last_step")
    captured_inputs: Dict[str, str] = Field(default_factory=dict, alias="captured_inputs")
    last_executed: datetime = Field(default_factory=datetime.utcnow, alias="last_executed")
    token_index: int = Field(default=0, alias="token_index")

    class Config:
        populate_by_name = True


class TokenState(BaseModel):
    current_index: int = Field(default=0, alias="current_index")
    history: Dict[str, BotExecutionState] = Field(default_factory=dict)

    class Config:
        populate_by_name = True


class TokenEntry:
    def __init__(self, token: str):
        self.token: str = token
        self.proxy: Optional[str] = None
        self.username: Optional[str] = None


class _TokenManager:
    """Singleton-like class untuk mengelola state."""

    def __init__(self):
        self._tokens: List[TokenEntry] = []
        self._proxy_list: List[str] = []
        self._state: TokenState = TokenState()
        self._owner: str = ""
        self._repo: str = ""
        self._token_cache: Dict[str, str] = {}
        self.http_timeout = httpx.Timeout(15.0, connect=5.0)

    def initialize(self):
        self._load_tokens()
        self._load_proxy_list()
        self._load_state()
        self._load_token_cache()
        self._assign_proxies_and_usernames()

    def reload_all_configs(self):
        console.print("[bold yellow]Reloading semua file konfigurasi...[/]")
        self._tokens.clear()
        self._proxy_list.clear()
        self._token_cache.clear()
        self._owner = ""
        self._repo = ""
        self._state = TokenState()  # Reset state
        self.initialize()
        console.print("[bold green]✓ Konfigurasi berhasil di-refresh.[/]")

    def _load_token_cache(self):
        if TOKEN_CACHE_PATH.exists():
            try:
                with open(TOKEN_CACHE_PATH, "r") as f:
                    self._token_cache = json.load(f)
                console.print(f"[green]Loaded {len(self._token_cache)} username dari cache[/]")
            except Exception as e:
                console.print(f"[yellow]Warning: Gagal memuat token cache: {e}[/]")
                self._token_cache = {}

    def save_token_cache(self, cache: Dict[str, str]):
        try:
            self._token_cache = cache
            with open(TOKEN_CACHE_PATH, "w") as f:
                json.dump(self._token_cache, f, indent=2)
        except Exception as e:
            console.print(f"[red]Error saving token cache: {e}[/]")

    def _load_tokens(self):
        if not TOKENS_PATH.exists():
            console.print(f"[red]ERROR: {TOKENS_PATH} tidak ditemukan![/]")
            console.print("[yellow]Buat file dengan format:[/]")
            console.print("[dim]Line 1: owner\nLine 2: repo\nLine 3: token1,token2,token3[/]")
            return

        try:
            with open(TOKENS_PATH, "r") as f:
                lines = f.readlines()
            
            if len(lines) < 3:
                console.print("[red]ERROR: github_tokens.txt format salah![/]")
                return

            self._owner = lines[0].strip()
            self._repo = lines[1].strip()
            
            tokens_str = lines[2].strip()
            self._tokens = [TokenEntry(t.strip()) for t in tokens_str.split(",") if t.strip()]
            
            console.print(f"[green]Loaded {len(self._tokens)} tokens untuk {self._owner}/{self._repo}[/]")

        except Exception as e:
            console.print(f"[red]Gagal membaca {TOKENS_PATH}: {e}[/]")

    def _load_proxy_list(self):
        if PROXY_LIST_PATH.exists():
            try:
                with open(PROXY_LIST_PATH, "r") as f:
                    self._proxy_list = [line.strip() for line in f if line.strip()]
                console.print(f"[green]Loaded {len(self._proxy_list)} proxy dari ProxySync[/]")
            except Exception as e:
                console.print(f"[yellow]Gagal membaca {PROXY_LIST_PATH}: {e}[/]")
        else:
            console.print("[yellow]ProxySync proxy.txt tidak ditemukan[/]")

    def _assign_proxies_and_usernames(self):
        if not self._tokens:
            return
        
        for i, entry in enumerate(self._tokens):
            if self._proxy_list:
                entry.proxy = self._proxy_list[i % len(self._proxy_list)]
            
            if entry.token in self._token_cache:
                entry.username = self._token_cache[entry.token]

    def _load_state(self):
        if STATE_PATH.exists():
            try:
                self._state = TokenState.model_validate_json(STATE_PATH.read_text())
            except Exception as e:
                console.print(f"[yellow]Warning: Gagal memuat state ({e}), membuat state baru.[/]")
                self._state = TokenState()
        else:
            self._state = TokenState()

    def _save_state(self):
        try:
            with open(STATE_PATH, "w") as f:
                f.write(self._state.model_dump_json(indent=2, by_alias=True))
        except Exception as e:
            console.print(f"[red]Error saving state: {e}[/]")

    def get_current_token_entry(self) -> Optional[TokenEntry]:
        if not self._tokens:
            return None
        return self._tokens[self._state.current_index]

    def get_repo_info(self) -> Tuple[str, str]:
        return self._owner, self._repo

    def get_all_token_entries(self) -> List[TokenEntry]:
        return self._tokens

    def get_username_cache(self) -> Dict[str, str]:
        return self._token_cache

    def switch_to_next_token(self):
        if not self._tokens:
            return
        
        self._state.current_index = (self._state.current_index + 1) % len(self._tokens)
        self._save_state()
        
        current = self.get_current_token_entry()
        if current:
            console.print(f"[yellow]Switch ke token #{self._state.current_index + 1} ({current.username or '???'})[/]")
            if current.proxy:
                console.print(f"[dim]Proxy: {current.proxy}[/]")

    def create_http_client(self, entry: Optional[TokenEntry] = None) -> httpx.AsyncClient:
        if entry is None:
            entry = self.get_current_token_entry()
            
        if entry is None:
            raise Exception("Tidak ada token yang dikonfigurasi")
            
        headers = {
            "Authorization": f"Bearer {entry.token}",
            "User-Agent": "Automation-Hub-Orchestrator/2.0-Python",
            "Accept": "application/vnd.github+json"
        }
        
        proxies = None
        if entry.proxy:
            # Asumsi format proxy: http://user:pass@host:port atau http://host:port
            if not entry.proxy.startswith(("http://", "https://", "socks5://")):
                proxy_url = f"http://{entry.proxy}"
            else:
                proxy_url = entry.proxy
            
            proxies = {
                "http://": proxy_url,
                "https://": proxy_url
            }
        
        return httpx.AsyncClient(headers=headers, proxies=proxies, timeout=self.http_timeout)

    def mask_token(self, token: str) -> str:
        return token[:10] + "..." + token[-7:] if len(token) > 20 else token

    def show_status(self):
        if not self._tokens:
            return
            
        table = Table(title="Status GitHub Token", expand=True)
        table.add_column("Index", style="dim")
        table.add_column("Token")
        table.add_column("Username")
        table.add_column("Proxy")
        table.add_column("Aktif", justify="center")

        for i, token in enumerate(self._tokens):
            is_active = "[green]✓[/]" if i == self._state.current_index else ""
            token_display = self.mask_token(token.token)
            
            proxy_display = token.proxy or "[yellow]no proxy[/]"
            if token.proxy and len(token.proxy) > 40:
                proxy_display = token.proxy[:37] + "..."

            table.add_row(
                str(i + 1),
                token_display,
                token.username or "[grey]???[/]",
                proxy_display,
                is_active
            )
        
        console.print(table)
        if self._proxy_list:
            console.print(f"\n[dim]Total proxy dari ProxySync: {len(self._proxy_list)}[/]")
        
        if self._state.history:
            console.print("\n[cyan]Histori Eksekusi:[/]")
            history_table = Table(expand=True)
            history_table.add_column("Bot")
            history_table.add_column("Last Step")
            history_table.add_column("Token Used")
            history_table.add_column("Last Run")
            
            sorted_history = sorted(
                self._state.history.items(), 
                key=lambda item: item[1].last_executed, 
                reverse=True
            )
            
            for bot_name, state in sorted_history:
                history_table.add_row(
                    bot_name,
                    str(state.last_step),
                    f"#{state.token_index + 1}",
                    state.last_executed.strftime("%Y-%m-%d %H:%M")
                )
            console.print(history_table)

# Inisialisasi instance singleton
TokenManager = _TokenManager()
